src/notify_telegram.c

```diff
@@ -0,0 +1,871 @@
+    }
+    char params[MAX_PARAMS_SIZE] = "";
+    snprintf(params, MAX_PARAMS_SIZE,
+             "url=%s:%s%s&allowed_updates=[\"message\"]", webhook_host,
+             webhook_port ? webhook_port : "443",
+             webhook_url ? webhook_url : "");
+    CURL *handle = curl_easy_init();
+    curl_easy_setopt(handle, CURLOPT_POSTFIELDS, params);
+    curl_easy_setopt(handle, CURLOPT_URL, url);
+    curl_easy_setopt(handle, CURLOPT_WRITEFUNCTION, tg_curl_write_callback);
+    curl_easy_setopt(handle, CURLOPT_WRITEDATA, (void *)&buf);
+    pthread_mutex_lock(&telegram_lock);
+    CURLcode response_code = curl_easy_perform(handle);
+    pthread_mutex_unlock(&telegram_lock);
+    curl_easy_cleanup(handle);
+    DEBUG("notify_telegram: curl response = %s", buf.response);
+    if (response_code != CURLE_OK) {
+      ERROR("notify_telegram: curl_easy_perform failed.");
+      sfree(buf.response);
+      return -1;
+    }
+  } else if (!webhook_host) {
+    DEBUG("notify_telegram: long polling started");
+
+    struct response_buffer_t buf = {0};
+    char url[MAX_URL_SIZE] = "";
+    if (proxy_url) {
+      snprintf(url, MAX_URL_SIZE, "%s%s/deleteWebhook", proxy_url, bot_token);
+    } else {
+      snprintf(url, MAX_URL_SIZE,
+               "https://api.telegram.org/bot%s/deleteWebhook", bot_token);
+    }
+    char params[MAX_PARAMS_SIZE] = "";
+    CURL *handle = curl_easy_init();
+    curl_easy_setopt(handle, CURLOPT_POSTFIELDS, params);
+    curl_easy_setopt(handle, CURLOPT_URL, url);
+    curl_easy_setopt(handle, CURLOPT_WRITEFUNCTION, tg_curl_write_callback);
+    curl_easy_setopt(handle, CURLOPT_WRITEDATA, (void *)&buf);
+    pthread_mutex_lock(&telegram_lock);
+    CURLcode response_code = curl_easy_perform(handle);
+    pthread_mutex_unlock(&telegram_lock);
+    curl_easy_cleanup(handle);
+    DEBUG("notify_telegram: curl response = %s", buf.response);
+    if (response_code != CURLE_OK) {
+      ERROR("notify_telegram: curl_easy_perform failed.");
+      sfree(buf.response);
+      return -1;
+    }
+  }
+
+  return notify_telegram_read();
+}
+
+static int notify_telegram_shutdown(void) {
+  curl_global_cleanup();
+
+  if (httpd) {
+    MHD_stop_daemon(httpd);
+    httpd = NULL;
+  }
+
+  return 0;
+}
+
+static int notify_telegram_config(const char *key, const char *value) {
+  if (strcasecmp(key, "RecipientChatID") == 0) {
+    char **tmp;
+    tmp = realloc(recipients, (recipients_len + 1) * sizeof(char *));
+    if (tmp == NULL) {
+      ERROR("notify_telegram: realloc failed.");
+      return -1;
+    }
+    recipients = tmp;
+    recipients[recipients_len] = strdup(value);
+    if (recipients[recipients_len] == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
+    recipients_len++;
+  } else if (strcasecmp(key, "BotToken") == 0) {
+    sfree(bot_token);
+    bot_token = strdup(value);
+    if (bot_token == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
```

Comment:
Code would be easier to read with a table containing the string and variable (address) pair, and a loop handling all of those.

Reply:
I don't think it's worth it, because not all keys in the config are handled the same way

Reply:
Most keys are, but indeed there would be few that would need separate handling outside of the proposed loop.  I'll resolve this.